File: 3834-merge-adjacent-equal-elements/3834-merge-adjacent-equal-elements.java
class Solution {
    public List<Long> mergeAdjacent(int[] nums) {

        Stack<Long> st = new Stack<>();

        for (int x : nums) {
            long num = x;   // 🔑 promote to long immediately

            while (!st.isEmpty() && st.peek() == num) {
                num += st.pop();   // safe now
            }

            st.push(num);
        }

        List<Long> arr = new ArrayList<>();
        while (!st.isEmpty()) {
            arr.add(st.pop());
        }

        Collections.reverse(arr);
        return arr;
    }
}

File: 3834-merge-adjacent-equal-elements/3834-merge-adjacent-equal-elements.py
class Solution(object):
    def mergeAdjacent(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        stack = []
        
        for num in nums:
            # While the current number matches the top of the stack,
            # they must be merged.
            while stack and stack[-1] == num:
                # Remove the top element and add it to our current value
                num += stack.pop()
            
            # Once no more merges are possible for this value, 
            # push the result onto the stack.
            stack.append(num)
            
        return stack